Implement Cleanup for Generator States in CUDAGraph Reset Method

alpha0422 opened the pull request:
Fixes #171263.


cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

alpha0422 committed:
Implement cleanup for generator states in CUDAGraph reset method.

Signed-off-by: Wil Kong <alpha0422@gmail.com>

galv commented on aten/src/ATen/cuda/CUDAGraph.cpp:
```diff
@@ -261,6 +261,35 @@ void CUDAGraph::reset() {
+    }
+    // Also end the allocator pool allocation if it was started.
+    // Use try-catch since reset() may be called from destructor and must not throw.
+    if (capture_dev_ != UNDEFINED_DEVICE && (mempool_id_.first != 0 || mempool_id_.second != 0)) {
```
You should be able to assert that `!(mempool_id_.first == 0 && mempool_id_.second == 0)` here. Either do that or remove this check. I'm also skeptical about whether capture_dev_ will ever be UNDEFINED_DEVICE.

galv commented on aten/src/ATen/cuda/CUDAGraph.cpp:
```diff
@@ -261,6 +261,35 @@ void CUDAGraph::reset() {
+
+  // If capture_begin() was called but capture_end() failed (e.g., due to
+  // sync operations like .item() during capture), the generator states
+  // will have capturing_ = true but no actual CUDA stream capture is active.
```
The problem is that the destructor of CUDAGraph will essentially always run after an exception like this is thrown. We don't want to have spurious errors that confuse users.

galv commented:
For what it's worth, I believe that there are other errors like this lurking. I've found them while working on this code. Unfortunately, they haven't ever been tackled in a coherent way because it doesn't often come up.

Can you please add a unit test?

Finally, please don't be upset if I decide to rewrite some of this. You've unearthed an issue that I worry may require some more in-depth analysis. The fact that the RNG states maintain their own capturing_ variable is a bit of a code smell imo.

alpha0422 committed:
Enhance CUDAGraph reset method.

- Ensure proper cleanup of generator states and memory pool after failed captures.
- Added unit test to verify recovery from capture failures, ensuring that RNG operations and allocated memory are correctly handled.

This addresses potential memory leaks and RNG errors following unsuccessful graph captures.

Signed-off-by: Wil Kong <alpha0422@gmail.com>

alpha0422 commented:
> For what it's worth, I believe that there are other errors like this lurking. I've found them while working on this code. Unfortunately, they haven't ever been tackled in a coherent way because it doesn't often come up.
> 
> Can you please add a unit test?
> 
> Finally, please don't be upset if I decide to rewrite some of this. You've unearthed an issue that I worry may require some more in-depth analysis. The fact that the RNG states maintain their own capturing_ variable is a bit of a code smell imo.

Added a unittest for the change. Feel free to rewrite and improve, thanks.

alpha0422 committed:
Fix issue reported by linter.

Signed-off-by: Wil Kong <alpha0422@gmail.com>

alpha0422 committed:
Move CUDAGraph cleanup on failure from reset to capture end.

Signed-off-by: Wil Kong <alpha0422@gmail.com>

alpha0422 commented:
> The diagnosis in the issue seems fine, but capturing the problem the next time we try to capture... feels bad. I kind of wish we just had a C++ try-catch to handle unsetting capturing upon exception...

That makes sense. I just moved the cleanup logic from reset to the end of capturing.

ezyang reviewed (changes requested):
You don't actually need a try catch, since the exception comes from AT_CUDA_CHECK, so you can just check the error code directly.

ezyang commented:
@claude Even with my comment above, the code seems kinda grungy. Any ideas what to do?
